sktime/dists_kernels/_base.py

```diff
@@ -0,0 +1,261 @@
+# -*- coding: utf-8 -*-
+"""
+Abstract base class for pairwise transformers (such as distance/kernel matrix makers)
+"""
+
+__author__ = ["fkiraly"]
+
+import numpy as np
+import pandas as pd
+
+from sktime.base import BaseEstimator
+
+from sktime.utils.validation.series import check_series
+
+
+class BasePairwiseTransformer(BaseEstimator):
+    """Base pairwise transformer for tabular or series data template class.
+
+    The base pairwise transformer specifies the methods and method
+    signatures that all pairwise transformers have to implement.
+
+    Specific implementations of these methods is deferred to concrete classes.
+    """
+
+    # default tag values - these typically make the "safest" assumption
+    _tags = {
+        "symmetric": False,  # is the transformer symmetric, i.e., t(x,y)=t(y,x) always?
+    }
+
+    def __call__(self, X, X2=None):
+        """
+        Behaviour: returns pairwise distance/kernel matrix
+            between samples in X and X2
+                if X2 is not passed, is equal to X
+
+        alias for transform
+
+        Parameters
+        ----------
+        X: pd.DataFrame of length n, or 2D np.array with n rows
+        X2: pd.DataFrame of length m, or 2D np.array with m rows, optional
+            default X2 = X
+
+        Returns
+        -------
+        distmat: np.array of shape [n, m]
+            (i,j)-th entry contains distance/kernel between X.iloc[i] and X2.iloc[j]
+
+        Writes to self
+        --------------
+        X_equals_X2: bool = True if X2 was not passed, False if X2 was passed
+            for use to make internal calculations efficient, e.g., in _transform
+        """
+        # no input checks or input logic here, these are done in transform
+        # this just defines __call__ as an alias for transform
+        return self.transform(X=X, X2=X2)
+
+    def transform(self, X, X2=None):
+        """
+        Behaviour: returns pairwise distance/kernel matrix
+            between samples in X and X2 (equal to X if not passed)
+
+        Parameters
+        ----------
+        X: pd.DataFrame of length n, or 2D np.array with n rows
+        X2: pd.DataFrame of length m, or 2D np.array with m rows, optional
+            default X2 = X
+
+        Returns
+        -------
+        distmat: np.array of shape [n, m]
+            (i,j)-th entry contains distance/kernel between X.iloc[i] and X2.iloc[j]
+
+        Writes to self
+        --------------
+        X_equals_X2: bool = True if X2 was not passed, False if X2 was passed
+            for use to make internal calculations efficient, e.g., in _transform
+        """
+
+        X = check_series(X)
+
+        if X2 is None:
+            X2 = X
+            self.X_equals_X2 = True
+        else:
+            X2 = check_series(X2)
+            self.X_equals_X2 = False
```

Comment:
I think X can be equal to X2 if X2 is passed, e.g. user may do `.transform(X,X)`

I recommend changing to something like `self.X_equals_X2 = (X == X2)`. (I don't think this is correct syntax)